### DISCUSSION TITLE
Howto set Callback-Handler-Function  to an no static class function

### DISCUSSION THREAD

**Discussion by LudwigWen** [eModbus/eModbus#124] [General]
please help:
 Howto set Callback-Handler-Function  within my own class xxx,  **without** declaring this function as STATIC
```
class xxx
public: 
    void xxx {
      **MB.onDataHandler(  ??? callbackData ??? ) ;** 
    }
    void callbackData(ModbusMessage response, uint32_t token) {   ...    }
private:
   ModbusClientTCPasync MB;
};
```

**Comment by Miq1** [comment]
This may be difficult - I never tried it myself.
The ``onDataHandler`` call will take a function pointer, so I would try with ``MB.onDataHandler(this->callbackData);``.

**Comment by bertmelis** [comment]
A non-static class method has a "hidden" argument that contains the pointer to the class instance but that violates the callback's signature.

However, because the callback can be a std::function it also accepts `std::bind`:

```cpp
// placeholder 1 for ModbusMessage and placeholder 2 for token
MB.onDataHandler(std::bind(YourClass::method, this, std::placeholders::_1, std::placeholders::_2));
```
